From be2690f8e355fd14c37dd934dd09d13fb9819902 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:39:31 -0400
Subject: [virt] kvm: Disable SMAP for guests in EPT realmode and EPT unpaging mode

Message-id: <1406040016-3289-177-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86255
O-Subject: [RHEL7 PATCH v2 176/221] KVM: Disable SMAP for guests in EPT realmode and EPT unpaging mode
Bugzilla: 1074747 1084616 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Feng Wu <feng.wu@intel.com>

SMAP is disabled if CPU is in non-paging mode in hardware.
However KVM always uses paging mode to emulate guest non-paging
mode with TDP. To emulate this behavior, SMAP needs to be
manually disabled when guest switches to non-paging mode.

Signed-off-by: Feng Wu <feng.wu@intel.com>
Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
(cherry picked from commit e1e746b3c55d1d1e0841a2e600a154d8f6747232)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 1320e0f..1f68c58 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -3484,13 +3484,14 @@ static int vmx_set_cr4(struct kvm_vcpu *vcpu, unsigned long cr4)
    hw_cr4 &= ~X86_CR4_PAE;
    hw_cr4 |= X86_CR4_PSE;
    /*
-    * SMEP is disabled if CPU is in non-paging mode in
-    * hardware. However KVM always uses paging mode to
+    * SMEP/SMAP is disabled if CPU is in non-paging mode
+    * in hardware. However KVM always uses paging mode to
     * emulate guest non-paging mode with TDP.
-    * To emulate this behavior, SMEP needs to be manually
-    * disabled when guest switches to non-paging mode.
+    * To emulate this behavior, SMEP/SMAP needs to be
+    * manually disabled when guest switches to non-paging
+    * mode.
     */
-   hw_cr4 &= ~X86_CR4_SMEP;
+   hw_cr4 &= ~(X86_CR4_SMEP | X86_CR4_SMAP);
   } else if (!(cr4 & X86_CR4_PAE)) {
    hw_cr4 &= ~X86_CR4_PAE;
   }
-- 
1.7.1